GoogleForCreators/web-stories-wp
# Quick Actions: Add Replace Media Functionality

## Context

People want to replace media, but _for real_

## Summary

Adds `Replace Media` functionality for foreground and background images.

## Relevant Technical Choices

The [`MediaUpload` component](https://github.com/google/web-stories-wp/blob/main/packages/wp-story-editor/src/components/mediaUpload/index.js#L26-L30) requires a render prop.

However, the `useQuickActions` hook does not render the quick action button. It's easier to pass in a click event handler that is later passed to the context menu button.

To do this, the `useMediaPicker` hook was exposed to the app. That way we don't need to use the render prop.

## To-do

- [x] Add replace functionality to background media
- [x] Add karma tests

## User-facing changes

|Replace foreground media|Replace background media|
|--|--|
|![replace-foreground](https://user-images.githubusercontent.com/22185279/138915789-b6bc8104-53cb-4009-a11f-1eb31e40fc51.gif)|![replace-background-media](https://user-images.githubusercontent.com/22185279/138915858-d99fc06e-a492-461c-bcd0-126302871425.gif)|


## Testing Instructions






This PR can be tested by following these steps:

**Make sure you do these for foreground media + background media (images, gifs, and videos)
1. Add a piece of media
2. Add styles to the media in the design panel
3. Click the `Replace` action in the quick action menu
4. Add a media and click `insert media`
5. New media should replace existing media but the styles should remain the same
6. Try the same thing by inserting media of different types (image, video, gif)

_Users without the `hasMediaUpload` capability should not be able to see this quick action_ (contributors)

## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #9196
